crates/runtime/src/accelerated_table/timestamp_metrics_utils.rs

```diff
@@ -0,0 +1,655 @@
+    ($ty:ty, $max_ts:expr, $array:expr, $value:expr) => {{
+        match $array.as_any().downcast_ref::<$ty>() {
+            Some(arr) => {
+                let batch_max = (0..arr.len())
+                    .filter(|&i| !arr.is_null(i))
+                    .filter_map(|i| $value(arr, i))
+                    .max();
+                match batch_max {
+                    Some(bm) => Some($max_ts.map_or(bm, |cur| cur.max(bm))),
+                    None => $max_ts,
+                }
+            }
+            None => $max_ts,
+        }
+    }};
+}
+
+/// Extracts the maximum timestamp from a stream of record batches.
+///
+/// The extraction uses `time_column` and `time_format` to interpret timestamps
+/// from various data types (timestamps, strings, integers, floats, dates).
+///
+/// # Parameters
+/// - `time_column`: Name of the column containing timestamp data
+/// - `time_format`: Format specification for interpreting the timestamp values
+/// - `source_name`: Used for logging warnings
+///
+/// # Returns
+/// - An updated data stream
+/// - An `Option<Arc<Mutex<Option<i64>>>>` with the following semantics:
+///     - `None`: Did not attempt to find a max timestamp (e.g., no `time_column`
+///       provided or column does not exist in schema)
+///     - `Some(Arc<Mutex<Option<i64>>>)`: Will be populated after the stream is consumed:
+///         - `Some(i64)`: The maximum timestamp found in the stream (in milliseconds)
+///         - `None`: Attempted extraction but no valid timestamps found
+///             - If batches were not empty, a warning is logged
+#[allow(clippy::too_many_lines)]
+pub async fn max_timestamp_in_stream(
+    data_update: StreamingDataUpdate,
+    schema: SchemaRef,
+    time_column: Option<String>,
+    time_format: Option<TimeFormat>,
+    source_name: String,
+) -> (StreamingDataUpdate, Option<Arc<Mutex<Option<i64>>>>) {
+    let Some(time_column) = time_column else {
+        return (data_update, None);
+    };
+
+    let time_format = time_format.unwrap_or_default();
+    let Some(field) = schema
+        .column_with_name(&time_column)
+        .map(|(_, f)| f)
+        .cloned()
+    else {
+        tracing::warn!(
+            "Failed to extract max_timestamp after refresh for {}: column {} not found in schema.",
+            source_name,
+            time_column,
+        );
+        return (data_update, None);
+    };
+
+    let max_ts = Arc::new(Mutex::new(None::<i64>));
+    let max_ts_clone = Arc::clone(&max_ts);
+
+    let out_stream = stream! {
+        let mut input_stream = data_update.data;
+        let mut max_ts: Option<i64> = None;
+
+        while let Some(batch_result) = input_stream.next().await {
+            if let Ok(ref batch) = batch_result
+                && let Ok(idx) = batch.schema().index_of(field.name()) {
+                    let array = batch.column(idx);
+
+                    let mut matched_supported_type = true;
+
+                    match field.data_type() {
```

Comment:
Can we move this match logic to a helper function to have main streaming logic more compact?